File: pkg/earthsci-ast-ts/src/assertion-nonfinite-scope.test.ts
/**
 * TypeScript's read of the SHARED `assertion_nonfinite` conformance manifest
 * (CONFORMANCE_SPEC §5.19, tests/conformance/assertion_nonfinite/manifest.json).
 *
 * TypeScript cannot run that category: the contract is the esm-spec §6.6.3 pass
 * predicate for a non-finite ACTUAL, and this binding has no simulator and no
 * assertion comparison at all — it parses a `tests` block as data and never
 * evaluates one. The manifest says so in `scope_excluded`, and this file is what
 * keeps that claim honest.
 *
 * Why a test rather than a line in a document. A shared corpus that only some
 * bindings read is a failure this suite has already had once: a rejection corpus
 * was being consumed by two of five bindings, and the three that ignored it
 * could have diverged silently for as long as nobody looked. An exclusion is the
 * same hazard one level down — invisible by construction. So all five bindings
 * read this manifest, and the two that cannot execute it assert their own
 * exclusion: if someone drops TypeScript from `scope_excluded` (or adds it to
 * `bindings_required`) without giving it a runner, this goes red instead of the
 * category quietly covering one binding fewer than it claims.
 */
import * as fs from 'node:fs'
import * as path from 'node:path'
import { describe, it, expect } from 'vitest'
import { loadString } from './parse.js'
import { fixturesDir } from './test-helpers.js'

interface NonfiniteCase {
  assertion_idx: number
  variable: string
  actual_class: '+inf' | '-inf' | 'nan' | 'finite'
  passed: boolean
}

interface NonfiniteManifest {
  category: string
  reference_binding: string
  bindings_required: string[]
  scope_excluded: Record<string, string>
  fixtures: Array<{ id: string; path: string; model: string; cases: NonfiniteCase[] }>
}

const CATEGORY_DIR = fixturesDir('conformance', 'assertion_nonfinite')

function manifest(): NonfiniteManifest {
  return JSON.parse(
    fs.readFileSync(path.join(CATEGORY_DIR, 'manifest.json'), 'utf8'),
  ) as NonfiniteManifest
}

describe('conformance: assertion_nonfinite (scope)', () => {
  it('excludes TypeScript by name, with a reason', () => {
    const m = manifest()
    expect(m.category).toBe('assertion_nonfinite')
    expect(m.reference_binding).toBe('julia')
    // Requiring TypeScript would be a claim this binding cannot honour: there
    // is nothing here that compares an assertion actual to an expectation.
    expect(m.bindings_required).not.toContain('typescript')
    expect(m.scope_excluded.typescript ?? '').not.toBe('')
  })

  it('carries a fixture this binding can still parse, and a non-vacuous case list', () => {
    const m = manifest()
    expect(m.fixtures.length).toBeGreaterThan(0)
    for (const fx of m.fixtures) {
      // A category whose fixture TypeScript cannot even load would be a format
      // divergence hiding behind a scope exclusion.
      const file = loadString(fs.readFileSync(path.join(CATEGORY_DIR, fx.path), 'utf8'))
      expect(Object.keys(file.models ?? {})).toContain(fx.model)

      // Non-vacuity, checkable without a simulator: at least one non-finite
      // case that MUST FAIL and at least one finite case that MUST PASS. A
      // manifest of nothing but failures would be satisfied by a binding that
      // failed every assertion.
      const mustFailNonFinite = fx.cases.filter((c) => c.actual_class !== 'finite' && !c.passed)
      const mustPassFinite = fx.cases.filter((c) => c.actual_class === 'finite' && c.passed)
      expect(mustFailNonFinite.length).toBeGreaterThan(0)
      expect(mustPassFinite.length).toBeGreaterThan(0)
    }
  })
})

File: pkg/earthsci-ast-go/pkg/esm/assertion_nonfinite_scope_test.go
package esm

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

// assertion_nonfinite_scope_test.go is Go's read of the SHARED
// `assertion_nonfinite` conformance manifest (CONFORMANCE_SPEC §5.20,
// tests/conformance/assertion_nonfinite/manifest.json).
//
// Go cannot run that category: the contract is the §6.6.3 pass predicate for a
// non-finite ACTUAL, and Go has no simulator and no assertion comparison — it
// parses a `tests` block as data and never evaluates one. The manifest says so
// in `scope_excluded`, and this test is what keeps that claim honest.
//
// Why a test rather than a line in a document. A shared corpus that only some
// bindings read is the failure this suite has already had once: a rejection
// corpus was being consumed by two of five bindings, and the three that ignored
// it could have diverged silently for as long as nobody looked. An exclusion is
// the same hazard one level down — it is invisible by construction. So every
// binding reads this manifest, and the two that cannot execute it assert their
// own exclusion: if someone drops Go from `scope_excluded` (or adds it to
// `bindings_required`) without giving Go a runner, this goes red instead of the
// category quietly covering one binding fewer than it claims.
func TestAssertionNonfiniteExcludesGoWithAReason(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}
	path := filepath.Join(repoRoot, "tests", "conformance", "assertion_nonfinite", "manifest.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}

	var manifest struct {
		Category         string            `json:"category"`
		ReferenceBinding string            `json:"reference_binding"`
		BindingsRequired []string          `json:"bindings_required"`
		ScopeExcluded    map[string]string `json:"scope_excluded"`
		Fixtures         []struct {
			Path  string `json:"path"`
			Cases []struct {
				AssertionIdx int    `json:"assertion_idx"`
				ActualClass  string `json:"actual_class"`
				Passed       bool   `json:"passed"`
			} `json:"cases"`
		} `json:"fixtures"`
	}
	if err := json.Unmarshal(raw, &manifest); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}

	if manifest.Category != "assertion_nonfinite" {
		t.Fatalf("category = %q, want assertion_nonfinite", manifest.Category)
	}
	for _, b := range manifest.BindingsRequired {
		if b == "go" {
			t.Fatal("the manifest requires Go, but Go has no assertion comparison to gate; " +
				"either give Go an inline-test runner or keep it in scope_excluded")
		}
	}
	if reason, ok := manifest.ScopeExcluded["go"]; !ok || reason == "" {
		t.Fatal("the manifest must say, in scope_excluded, WHY Go does not run this category")
	}

	// The fixture the other three bindings run must be present and loadable as
	// an ESM document by this binding, even though Go cannot evaluate its
	// assertions: a category whose fixture Go cannot even parse would be a
	// format divergence hiding behind a scope exclusion.
	if len(manifest.Fixtures) == 0 {
		t.Fatal("the manifest declares no fixtures")
	}
	for _, fx := range manifest.Fixtures {
		fixturePath := filepath.Join(repoRoot, "tests", "conformance", "assertion_nonfinite", fx.Path)
		doc, err := LoadPath(fixturePath)
		if err != nil {
			t.Fatalf("load %s: %v", fixturePath, err)
		}
		if len(doc.Models) == 0 {
			t.Fatalf("%s declares no models", fixturePath)
		}
		if len(fx.Cases) == 0 {
			t.Fatalf("%s has no declared cases", fx.Path)
		}
		// Non-vacuity, checkable without a simulator: the category must carry
		// at least one non-finite case that MUST FAIL and at least one finite
		// case that MUST PASS. A manifest of nothing but failures would be
		// satisfiable by a binding that failed every assertion.
		var mustFailNonFinite, mustPassFinite int
		for _, c := range fx.Cases {
			if c.ActualClass != "finite" && !c.Passed {
				mustFailNonFinite++
			}
			if c.ActualClass == "finite" && c.Passed {
				mustPassFinite++
			}
		}
		if mustFailNonFinite == 0 || mustPassFinite == 0 {
			t.Fatalf("%s: %d must-fail non-finite and %d must-pass finite cases; "+
				"the category needs both to be non-vacuous", fx.Path, mustFailNonFinite, mustPassFinite)
		}
	}
}
